internals/overlord/state/state.go

```diff
@@ -570,6 +568,131 @@ NextChange:
 			delete(s.tasks, tid)
 		}
 	}
+	// Prune expired notices, and update the latest warning time cache.
+	var latestWarningTime time.Time
+	for k, n := range s.notices {
+		if n.expired(now) {
+			stats.IncludeNotice(n)
+			delete(s.notices, k)
+		} else if n.noticeType == WarningNotice {
+			if n.lastRepeated.After(latestWarningTime) {
+				latestWarningTime = n.lastRepeated
+			}
+		} else if n.noticeType == ChangeUpdateNotice {
+			if _, exists := s.changes[n.key]; !exists {
+				stats.IncludeNotice(n)
+				delete(s.notices, k)
+			}
+		}
+	}
+	s.latestWarningTime.Store(&latestWarningTime)
+	if len(s.notices) > maxNotices {
+		s.pruneMaxNotices(maxNotices, stats)
+	}
+	stats.Log()
+}
+
+// pruneStats tracks how many changes and notices have been pruned, and what
+// time range has been affected.
+type pruneStats struct {
+	numChangesPruned   int
+	numNoticesPruned   int
+	oldestChangePruned time.Time
+	newestChangePruned time.Time
+	oldestNoticePruned time.Time
+	newestNoticePruned time.Time
+}
+
+func (p *pruneStats) IncludeChange(chg *Change) {
+	p.numChangesPruned++
+	if chg == nil || chg.readyTime.IsZero() {
+		return
+	}
+	if p.oldestChangePruned.IsZero() || p.oldestChangePruned.After(chg.readyTime) {
+		p.oldestChangePruned = chg.readyTime
+	}
+	if p.newestChangePruned.IsZero() || p.newestChangePruned.Before(chg.readyTime) {
+		p.newestChangePruned = chg.readyTime
+	}
+}
+
+func (p *pruneStats) IncludeNotice(n *Notice) {
+	p.numNoticesPruned++
+	if n == nil || n.lastOccurred.IsZero() {
+		return
+	}
+	if p.oldestNoticePruned.IsZero() || p.oldestNoticePruned.After(n.lastOccurred) {
+		p.oldestNoticePruned = n.lastOccurred
+	}
+	if p.newestNoticePruned.IsZero() || p.newestNoticePruned.Before(n.lastOccurred) {
+		p.newestNoticePruned = n.lastOccurred
+	}
+}
+
+func (p *pruneStats) Log() {
```

Comment:
I think we should significantly simplify this (and don't log anything if they're both 0). Also nit: let's change the receiver name `p` to `s` ("stats" makes more sense when its used). How about:

```go
if s.numChanges > 0 {
    logger.Noticef("pruned %d changes from %s to %s", ...)
}
if s.numNotices > 0 {
    logger.Noticef("pruned %d notices from %s to %s", ...)
}
```

Reply:
I updated the receiver, but I'm not sure I agree about not reporting when they are both 0. I'm happy to drop it to Debug, but not knowing whether it is actually occurring or not is very relevant in logs.
I tweaked it a bit. Let me know what you think.